### DISCUSSION TITLE
Proper random number generation

### DISCUSSION THREAD

**Discussion by moodmosaic** [AutoFixture/AutoFixture#1040] [Q&A]
*This goes back to #2, one of the very first issues we've created here, on GitHub, after @ecampidoglio migrated the source code from CodePlex.* :rocket:

---

[RandomNumericSequenceGenerator](https://github.com/AutoFixture/AutoFixture/blob/e6293de7c8834d5dd42e1f81868af768d08f3438/Src/AutoFixture/RandomNumericSequenceGenerator.cs) is basically a wrapper on top of [System.Random](https://msdn.microsoft.com/en-us/library/system.random.aspx) with the addition of 64-bit integers and no duplicates (using an, inefficient, `HashSet<T>` internally for this).

I was looking at the source code of RandomNumericSequenceGenerator and I was wondering whether it's *random enough*. Overtime, several people had [issues](https://github.com/AutoFixture/AutoFixture/issues?q=is%3Aissue+random+label%3Abug) with it.

## Keep using RandomNumericSequenceGenerator 

We could keep using RandomNumericSequenceGenerator as long as it's random enough... How do we find out?

There are several statistical test suites we can run, e.g. [Dieharder](https://webhome.phy.duke.edu/~rgb/General/dieharder.php), [TestU01](http://simul.iro.umontreal.ca/testu01/tu01.html), [PractRand](http://pracrand.sourceforge.net/). I've  used Dieharder [to test the PRNG we use in Hedgehog](https://github.com/hedgehogqa/haskell-hedgehog/pull/185). I'd be tempted to try PractRand after seeing Melissa E. O’Neill's [post](http://www.pcg-random.org/posts/how-to-test-with-practrand.html).

## Switch to a proper PRNG

There are several out there already, some of them are battle-tested.

* Donald E. Knuth: *Subtractive random number generator* ([paper](https://dl.acm.org/citation.cfm?id=270146)) — used in .NET [System.Random](https://msdn.microsoft.com/en-us/library/system.random.aspx)
* Pierre L'Ecuyer: *Efficient and portable combined random number generators* ([paper](https://www.iro.umontreal.ca/~lecuyer/myftp/papers/cacm88.pdf)) — used in Haskell [random](http://hackage.haskell.org/package/random) and [QuickCheck](http://hackage.haskell.org/package/QuickCheck), and FsCheck v1 and v2.
* Koen Claessen and Michał H. Pałka: [TFGen](http://publications.lib.chalmers.se/records/fulltext/183348/local_183348.pdf) — used in Haskell QuickCheck.
* Guy L. Steele, Doug Lea, Christine H. Flood: [SplitMix](http://dl.acm.org/citation.cfm?id=2660195) — used in the [Java SDK](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/java/util/SplittableRandom.java), [Haskell Hedgehog](https://github.com/hedgehogqa/haskell-hedgehog/blob/39b15b9b4d147f6001984c4b7edab00878269da7/hedgehog/src/Hedgehog/Internal/Seed.hs), [F# Hedgehog](https://github.com/hedgehogqa/fsharp-hedgehog/blob/bd57210c2593ecc8ba053ac76f9bfc3cee5407e9/src/Hedgehog/Seed.fs), and [soon in FsCheck v3](https://github.com/fscheck/FsCheck/issues/198#issuecomment-184136162).
* Melissa E. O’Neill: [PCG](http://www.pcg-random.org): *A Family of Simple Fast Space-Efficient Statistically Good Algorithms for Random Number Generation* ([paper](http://www.pcg-random.org/pdf/hmc-cs-2014-0905.pdf))

Implementations of the [Mersenne Twister](http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/emt.html) PRNG and the [FNV-1a](http://www.isthe.com/chongo/tech/comp/fnv/) hashing algorithm are used in C [theft](https://github.com/silentbicycle/theft), and I'm not currently aware what [hypothesis](https://github.com/HypothesisWorks/hypothesis/tree/master/hypothesis-python) uses.

### Implementation

As long as the implementation is small and minimal, e.g.

```c
// *Really* minimal PCG32 code / (c) 2014 M.E. O'Neill / pcg-random.org
// Licensed under Apache License 2.0 (NO WARRANTY, etc. see website)

typedef struct { uint64_t state;  uint64_t inc; } pcg32_random_t;

uint32_t pcg32_random_r(pcg32_random_t* rng)
{
    uint64_t oldstate = rng->state;
    // Advance internal state
    rng->state = oldstate * 6364136223846793005ULL + (rng->inc|1);
    // Calculate output function (XSH RR), uses old state for max ILP
    uint32_t xorshifted = ((oldstate >> 18u) ^ oldstate) >> 27u;
    uint32_t rot = oldstate >> 59u;
    return (xorshifted >> rot) | (xorshifted << ((-rot) & 31));
}
```

and passes Dieharder, PractRand, or TestU01, we should be good to go.

There's a bug in the paper of SplitMix, found by Melissa E. O’Neill, so we need to keep https://github.com/hedgehogqa/haskell-hedgehog/commit/39b15b9b4d147f6001984c4b7edab00878269da7 in mind and port it from Haskell Hedgehog or [Java](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/SplittableRandom.java) (if we decide to use SplitMix).

---

For all the public, and great, work Melissa E. O’Neill has been doing on [pcg-random.org](http://www.pcg-random.org), I lean towards PCG.

**Comment by ploeh** [comment]
I can't say that I have opinions about which RNG is used. It was never the intent that AutoFixture should be cryptographically random, and I don't see how that's important. I agree that it matters for libraries like Hedehog and FsCheck, but is it important in the context of AutoFixture?

Isn't [that issue list](https://github.com/AutoFixture/AutoFixture/issues?q=is%3Aissue+random+label%3Abug) a bit of a straw man? All but one of those issues are closed, and the open issue doesn't seem to be caused by AutoFixture's random value generator not being sufficiently random. I've only skimmed the issue, though, so I could be wrong.

My original intent with AutoFixture was to make some assumptions about how users' SUTs would typically look, and then try to generate values that would belong to the domain of the the objects involved, and stay within the same equivalence class across test runs.

As an example, when generating integers, I found it prudent to generate small, positive numbers. People often use integers to represent natural numbers, here with an emphasis on _natural_; the number of items in a shopping basket, the number of rows in a list on a screen, the number of days until something happens, and so on. Numbers like 11, 231, or 89 are often fine for testing most systems that operate on natural numbers, whereas -10, 184,819,099, or -987 aren't.

That concerns is, in my opinion, more important than whether the random value generator is truly random.

That said, I still have no opinion on which RNG is being used, as long as the above concern is taken into account, but since a strong preference for 'small' positive numbers inherently introduces a skew in the selection process, does it even matter?

**Comment by moodmosaic** [comment]
>It was never the intent that AutoFixture should be cryptographically random, and I don't see how that's important.

To be clear, none of those random number generators should be used for crypto or security applications. I only care about whether the current implementation is adequate for "everyday" use.—If so, it's all good.

**Comment by zvirja** [comment]
@moodmosaic I also don't have any opinion on that. For me the randomness is not the issue, but that could be bound to my personal experience. The major issue is that current implementation is not efficient due to `HashSet<>` and if we mitigate that, it would be awesome.

If change of the approach solves both the issues simultaneously, then I'm both hands in. But we should ensure the fulfill the progressive limits we currently have (like Mark mentioned above).

**Comment by moodmosaic** [comment]
IIRC, the `HashSet<T>` is there to prevent the same number from being returned more than once. But, this can be the responsibility of the consumer, to *filter* any unwanted numbers.

---

If you take the `HashSet<T>` concept out I don't see dramaticall differences from just using [Random.Next](https://msdn.microsoft.com/en-us/library/system.random.next.aspx) with an injected seed value in the Composition Root (the `Fixture` class).
